Extract markers into plugin

illetid opened the pull request:
**Type of PR:** enhancement

- `N/A` ~Addresses an existing issue: fixes #~
- [x] Includes tests
- [x] Documentation update

**Overview of change:**
Moved a series markers into a built-in plugin to enable tree shaking

SlicedSilver commented on src/index.ts:
```diff
@@ -26,6 +26,7 @@ export { createChart, createChartEx, defaultHorzScaleBehavior } from './api/crea
 */
 export { TextWatermark } from './plugins/text-watermark/primitive';
 export { ImageWatermark } from './plugins/image-watermark/primitive';
+export { createSeriesMarkersPrimitive } from './plugins/series-markers/primitive';
```
If we are going to use a `createSeriesMarkersPrimitive` function then I think it would be better if it returned an API interface instead of the actual primitive. At the moment the primitive exposes methods like `attached`, `detached`, `hitTest`, ... when it might make sense to only expose the `setMarkers` method.

So perhaps the best approach is to:
- export the primitive classes (so that people can extend if required)
- export `create...` functions for each plugin which adds the primitive, and provides a simplified wrapper API around the primitive which exposes only a few methods (like `setMarkers`, and `remove` which detaches the primitive).

SlicedSilver replied:
Having a look at the `d.ts` file, I'm not a fan of the Plugin classes having their private methods and fields displayed there (it's the same for the other plugins like the watermarks).

So I'm even more in favour of using `create` functions with a simplified public API wrapper around the primitive so that we can hide these implementation details. We are an open-source library so perhaps the benefit of exporting the classes so that they can be extended isn't really required because people can view the source anyway.

So what are your thoughts on no longer exporting the primitive classes for plugins and rather exposing a nice wrapper around them?
There could be a generic wrapper class which exposes some common things like: `detach` and maybe `applyOptions` and then each actual wrapper can extend that and add things like `setMarkers`, `markers` (to read), `series` (to get the attached series)...

I think these would be cleaner and less confusing.

illetid replied:
agree that exporting function instead of primitive is better from a user perspective, and if someone what to extend he can just copy from the sourcecode.
I tried to create a `create` functions, not sure about typings with this `WithOptions` stuff, could you take a looks pls. Also I updated ts to latests, because it didn't see an `applyOptions` methods even it worked in vscode, so bump version solved the issue.

SlicedSilver replied:
Thanks. I think the create functions are nicer 👍

SlicedSilver commented on src/index.ts:
```diff
@@ -26,6 +26,7 @@ export { createChart, createChartEx, defaultHorzScaleBehavior } from './api/crea
 */
 export { TextWatermark } from './plugins/text-watermark/primitive';
 export { ImageWatermark } from './plugins/image-watermark/primitive';
+export { createSeriesMarkersPrimitive } from './plugins/series-markers/primitive';
```
Could we add some type checks tests within the `tests/type-checks` folder?

For example: `series-markers.ts`
```ts
import { createChart, createSeriesMarkersPrimitive } from '../../src';
import { Time } from '../../src/model/horz-scale-behavior-time/types';

const chart = createChart('anything');

const mainSeries = chart.addLineSeries();
mainSeries.setData([]);

const seriesMarkersPrimitive = createSeriesMarkersPrimitive(mainSeries, [
	{
		color: 'green',
		position: 'inBar',
		shape: 'arrowDown',
		time: 1556880900 as Time,
	},
]);

seriesMarkersPrimitive.setMarkers([
	{
		color: 'red',
		position: 'aboveBar',
		shape: 'arrowDown',
		time: 1556880900 as Time,
	},
]);

// and also test with a non-time-based series chart
```

Just to make sure that people using the library in typescript aren't going to get type errors when using it correctly. (See the other examples).

Currently I'm getting an error about not including `originalTime` in the marker object. This is marked as `@internal` so it won't be in the actual release and won't be an error but I don't think we need `originalTime` anyway. It could be an internal type of the primitive / view / renderer. And secondly I think we could remove it completely anyway (I don't see it being used).

illetid replied:
added type-checks, thanks, I tested only on javascript examples and miss that

SlicedSilver commented on src/model/autoscale-info-impl.ts:
```diff
@@ -29,11 +29,8 @@ export class AutoscaleInfoImpl {
-		}
 		return {
-			priceRange: this._priceRange.toRaw(),
+			priceRange: this._priceRange === null ? null : this._priceRange.toRaw(),
```
This return type for `toRaw()` is no longer correct. It now always returns `AutoscaleInfo`.

SlicedSilver commented on src/plugins/series-markers/primitive.ts:
```diff
@@ -0,0 +1,229 @@
+} from './utils';
+
+class SeriesMarkersPrimitive<HorzScaleItem> implements ISeriesPrimitive<HorzScaleItem> {
+	private _attached: SeriesAttachedParameter<HorzScaleItem> | null = null;
```
I don't think we need to keep a reference to this. We only use it to access `this._attached.series` which is also available from `this._series`

illetid replied:
removed, thanks

SlicedSilver commented on src/plugins/series-markers/primitive.ts:
```diff
@@ -0,0 +1,229 @@
+		return this._autoScaleMargins;
+	}
+
+	protected _getMarkerPositions(): MarkerPositions {
```
Does this need to be `protected`? Seems like it could rather be marked as `private`

illetid replied:
changed

SlicedSilver commented on src/plugins/series-markers/types.ts:
```diff
@@ -11,7 +13,7 @@ export type SeriesMarkerShape = 'circle' | 'square' | 'arrowUp' | 'arrowDown';
  * Represents a series marker.
  */
-export interface SeriesMarker<TimeType> {
+export interface SeriesMarker<TimeType = Time> {
```
Do we need to specify the default for the TimeType generic here? Hopefully it would be correctly inferred.

illetid replied:
no, thanks for spotting this

SlicedSilver commented on src/plugins/series-markers/utils.ts:
```diff
@@ -48,3 +48,16 @@ export function calculateAdjustedMargin(margin: number, hasSide: boolean, hasInB
 
 	return 0;
 }
+export function convertSeriesMarker<InTimeType, OutTimeType>(sm: SeriesMarker<InTimeType>, newTime: OutTimeType, originalTime?: unknown): SeriesMarker<OutTimeType> {
```
Is this used?

illetid replied:
removed, thanks

SlicedSilver reviewed (commented):
- [x] Could we also add a short documentation page for the plugin?
- [x] I know the graphics tests will fail because the interface changed. Did you manage to run the tests locally before the changes, save those images somewhere, and then run again with the current changes? If so, could you check that the screenshots are the same? It's a bit of pain to manually do this I guess. Downside of our current screenshot testing system.

illetid commented on .size-limit.js:
```diff
@@ -8,31 +8,31 @@ export default [
 		name: 'ESM createChart',
 		path: 'dist/lightweight-charts.production.mjs',
-		limit: '45.00 KB',
+		limit: '41.30 KB',
```
reverted, thanks for letting know, nice idea, changing it every time was horrible

illetid replied:
and changed docs, we had a series markers page, so I just change the method calls there

SlicedSilver commented on src/plugins/pane-primitive-wrapper.ts:
```diff
@@ -0,0 +1,43 @@
+}
+
+export class PanePrimitiveWrapper<T, Options = unknown, TPrimitive extends IPanePrimitiveWithOptions<T, Options> = IPanePrimitive<T>> implements IPrimitiveWrapper<T, Options> {
+	protected _primitive: TPrimitive;
+	protected _pane: IPaneApi<T>;
```
Lets try use `private` where possible and only use `protected` if we need to.
```suggestion
	private _primitive: TPrimitive;
	private _pane: IPaneApi<T>;
```

SlicedSilver commented on src/plugins/image-watermark/primitive.ts:
````diff
@@ -109,3 +94,31 @@ export class ImageWatermark<T = Time> implements IPanePrimitive<T> {
+ * imageWatermark.detach();
+ * ```
+ */
+export function createImageWatermark<T>(pane: IPaneApi<T>, imageUrl: string, options: DeepPartial<ImageWatermarkOptions>): PanePrimitiveWrapper<T, ImageWatermarkOptions, IPanePrimitiveWithOptions<T, ImageWatermarkOptions>> {
````
[This applies to all the `create...()` functions]
We should rather define the public API for the plugin wrapper by defining an interface like `ITextWatermarkPluginApi` which defines only the public methods so that the generated `d.ts` typings doesn't include any `private` methods.
I'll send you an example code snippet.

illetid replied:
added suggested changes, now the typings looks more clean, thanks

SlicedSilver commented on website/docs/migrations/from-v4-to-v5.md:
```diff
@@ -13,16 +13,17 @@ In the new version of Lightweight Charts, the watermark feature has undergone si
 3. **Improved Tree-shaking**: This change makes the feature more tree-shakeable, potentially reducing bundle sizes for users who don't need watermarks.
 4. **Added an Image Watermark Primitive**: In addition to the usual text based watermark, there is now
-an image watermark feature provided by the [`ImageWatermark`](/api/classes/ImageWatermark.md) primitive.
+an image watermark feature provided by the [`createImageWatermark`](/api/functions/createImageWatermark.md) primitive.
```
Lets not call it a primitive here, maybe call it a plugin?

SlicedSilver commented on tests/type-checks/series-markers.ts:
```diff
@@ -0,0 +1,122 @@
+const lineSeries = nonDefaultChart.addLineSeries();
+lineSeries.setData([]);
+
+const markerApi = createSeriesMarkers(mainSeries, [
+	{
+		color: 'green',
+		position: 'inBar',
+		shape: 'arrowDown',
+		time: 1556880900 as Time,
+	},
+]);
+
+markerApi.setMarkers([
+	{
+		color: 'red',
+		position: 'aboveBar',
+		shape: 'arrowDown',
+		time: 1556880900 as Time,
+	},
+]);
```
Using the wrong series here...
```suggestion
const timeValue: HorizontalScaleType = 12345;
const markerApi = createSeriesMarkers(lineSeries, [
	{
		color: 'green',
		position: 'inBar',
		shape: 'arrowDown',
		time: timeValue,
	},
]);

markerApi.setMarkers([
	{
		color: 'red',
		position: 'aboveBar',
		shape: 'arrowDown',
		time: 1556880900,
	},
]);
```

SlicedSilver commented on tests/type-checks/non-time-based-custom-series.ts:
```diff
@@ -107,7 +107,6 @@ if (dataSet) {
 dataSet.push({ time: 12 });
 
-const textWatermark = new TextWatermark<HorizontalScaleType>({
+createTextWatermark<HorizontalScaleType>(chart.panes()[1], {
```
Lets have the test check that we don't need to specify the generic because it should be automatically inferred. We can use a `satisfies` to then ensure that the type is expected:
```ts
createTextWatermark(chart.panes()[1], {
	lines: [],
}) satisfies ITextWatermarkPluginApi<number>;

createTextWatermark(chart.panes()[1], {
	lines: [],
// @ts-expect-error Time is not the expected Generic here.
}) satisfies ITextWatermarkPluginApi<Time>;
```

SlicedSilver commented:
- There is a failing test within the 'coverage' e2e suite.
- Could you check the graphics e2e tests and ensure that all the tests are generating a `2. test.png` and that the image looks correct. You might need to run the tests locally on golden (v5-candidate) with these line removed: https://github.com/tradingview/lightweight-charts/blob/a20f0846139d10b0752b0d3be26a313a7ec78d2b/tests/e2e/graphics/graphics-test-cases.ts#L213 to get the correct baseline images and store them somewhere. Then run the e2e tests again but this time on this branch (and that line added back in), and then manually compare the images. Perhaps by having a script get the matching images for the baseline folder and the tests folder → place them into a new directory with naming like: `series-markers_test.png` and `series-markers_golden.png` and then you can quickly go through the images one by one and quickly notice if anything looks different. (mac quicklook and arrow keys work well for this)

illetid commented:
> * There is a failing test within the 'coverage' e2e suite.
> * Could you check the graphics e2e tests and ensure that all the tests are generating a `2. test.png` and that the image looks correct. You might need to run the tests locally on golden (v5-candidate) with these line removed: https://github.com/tradingview/lightweight-charts/blob/a20f0846139d10b0752b0d3be26a313a7ec78d2b/tests/e2e/graphics/graphics-test-cases.ts#L213
>    to get the correct baseline images and store them somewhere. Then run the e2e tests again but this time on this branch (and that line added back in), and then manually compare the images. Perhaps by having a script get the matching images for the baseline folder and the tests folder → place them into a new directory with naming like: `series-markers_test.png` and `series-markers_golden.png` and then you can quickly go through the images one by one and quickly notice if anything looks different. (mac quicklook and arrow keys work well for this)

I've checked all the screenshots. Fixed one small issue with margins that I've spotted in progress, and failing coverage tests are also fixed, I used wrong interface as returned type and ts-transformer renamed properties that should be public.
attached zip with tests results and added in each folder image from reference run to check with 
[devicePixelRatio=2.00.zip](https://github.com/user-attachments/files/16869847/devicePixelRatio.2.00.zip)

SlicedSilver commented on src/plugins/series-markers/primitive.ts:
```diff
@@ -0,0 +1,189 @@
+		return null;
+	}
+
+	public autoScaleMargins(): AutoScaleMargins | null {
```
Could we make this `private`?

illetid replied:
changed, thanks

SlicedSilver reviewed (approved):
:shipit:
